File: ruby/earthquakes.rb
# This code may look unusually verbose for Ruby (and it is), but
# it performs some subtle and complex validation of JSON data.
#
# To parse this JSON, add 'dry-struct' and 'dry-types' gems, then do:
#
#   earthquakes = Earthquakes.from_json! "{…}"
#   puts earthquakes.bbox.first
#
# If from_json! succeeds, the value returned matches the schema.

require 'json'
require 'dry-types'
require 'dry-struct'

module Types
  include Dry::Types.module

  Int    = Strict::Int
  Nil    = Strict::Nil
  Hash   = Strict::Hash
  String = Strict::String
  Double = Strict::Float | Strict::Int
end

class Geometry < Dry::Struct
  attribute :geometry_type, Types::String
  attribute :coordinates,   Types.Array(Types::Double)

  def self.from_dynamic!(d)
    d = Types::Hash[d]
    new(
      geometry_type: d.fetch("type"),
      coordinates:   d.fetch("coordinates"),
    )
  end

  def self.from_json!(json)
    from_dynamic!(JSON.parse(json))
  end

  def to_dynamic
    {
      "type"        => @geometry_type,
      "coordinates" => @coordinates,
    }
  end

  def to_json(options = nil)
    JSON.generate(to_dynamic, options)
  end
end

class Properties < Dry::Struct
  attribute :mag,             Types::Double
  attribute :place,           Types::String
  attribute :time,            Types::Int
  attribute :updated,         Types::Int
  attribute :tz,              Types::Int
  attribute :url,             Types::String
  attribute :detail,          Types::String
  attribute :felt,            Types::Int.optional
  attribute :cdi,             Types::Int.optional
  attribute :mmi,             Types::Double.optional
  attribute :alert,           Types::Nil
  attribute :status,          Types::String
  attribute :tsunami,         Types::Int
  attribute :sig,             Types::Int
  attribute :net,             Types::String
  attribute :code,            Types::String
  attribute :ids,             Types::String
  attribute :sources,         Types::String
  attribute :types,           Types::String
  attribute :nst,             Types::Int.optional
  attribute :dmin,            Types::Double.optional
  attribute :rms,             Types::Double
  attribute :gap,             Types::Int.optional
  attribute :mag_type,        Types::String
  attribute :properties_type, Types::String
  attribute :title,           Types::String

  def self.from_dynamic!(d)
    d = Types::Hash[d]
    new(
      mag:             d.fetch("mag"),
      place:           d.fetch("place"),
      time:            d.fetch("time"),
      updated:         d.fetch("updated"),
      tz:              d.fetch("tz"),
      url:             d.fetch("url"),
      detail:          d.fetch("detail"),
      felt:            d.fetch("felt"),
      cdi:             d.fetch("cdi"),
      mmi:             d.fetch("mmi"),
      alert:           d.fetch("alert"),
      status:          d.fetch("status"),
      tsunami:         d.fetch("tsunami"),
      sig:             d.fetch("sig"),
      net:             d.fetch("net"),
      code:            d.fetch("code"),
      ids:             d.fetch("ids"),
      sources:         d.fetch("sources"),
      types:           d.fetch("types"),
      nst:             d.fetch("nst"),
      dmin:            d.fetch("dmin"),
      rms:             d.fetch("rms"),
      gap:             d.fetch("gap"),
      mag_type:        d.fetch("magType"),
      properties_type: d.fetch("type"),
      title:           d.fetch("title"),
    )
  end

  def self.from_json!(json)
    from_dynamic!(JSON.parse(json))
  end

  def to_dynamic
    {
      "mag"     => @mag,
      "place"   => @place,
      "time"    => @time,
      "updated" => @updated,
      "tz"      => @tz,
      "url"     => @url,
      "detail"  => @detail,
      "felt"    => @felt,
      "cdi"     => @cdi,
      "mmi"     => @mmi,
      "alert"   => @alert,
      "status"  => @status,
      "tsunami" => @tsunami,
      "sig"     => @sig,
      "net"     => @net,
      "code"    => @code,
      "ids"     => @ids,
      "sources" => @sources,
      "types"   => @types,
      "nst"     => @nst,
      "dmin"    => @dmin,
      "rms"     => @rms,
      "gap"     => @gap,
      "magType" => @mag_type,
      "type"    => @properties_type,
      "title"   => @title,
    }
  end

  def to_json(options = nil)
    JSON.generate(to_dynamic, options)
  end
end

class Feature < Dry::Struct
  attribute :feature_type, Types::String
  attribute :properties,   Properties
  attribute :geometry,     Geometry
  attribute :id,           Types::String

  def self.from_dynamic!(d)
    d = Types::Hash[d]
    new(
      feature_type: d.fetch("type"),
      properties:   Properties.from_dynamic!(d.fetch("properties")),
      geometry:     Geometry.from_dynamic!(d.fetch("geometry")),
      id:           d.fetch("id"),
    )
  end

  def self.from_json!(json)
    from_dynamic!(JSON.parse(json))
  end

  def to_dynamic
    {
      "type"       => @feature_type,
      "properties" => @properties.to_dynamic,
      "geometry"   => @geometry.to_dynamic,
      "id"         => @id,
    }
  end

  def to_json(options = nil)
    JSON.generate(to_dynamic, options)
  end
end

class Metadata < Dry::Struct
  attribute :generated, Types::Int
  attribute :url,       Types::String
  attribute :title,     Types::String
  attribute :status,    Types::Int
  attribute :api,       Types::String
  attribute :count,     Types::Int

  def self.from_dynamic!(d)
    d = Types::Hash[d]
    new(
      generated: d.fetch("generated"),
      url:       d.fetch("url"),
      title:     d.fetch("title"),
      status:    d.fetch("status"),
      api:       d.fetch("api"),
      count:     d.fetch("count"),
    )
  end

  def self.from_json!(json)
    from_dynamic!(JSON.parse(json))
  end

  def to_dynamic
    {
      "generated" => @generated,
      "url"       => @url,
      "title"     => @title,
      "status"    => @status,
      "api"       => @api,
      "count"     => @count,
    }
  end

  def to_json(options = nil)
    JSON.generate(to_dynamic, options)
  end
end

class Earthquakes < Dry::Struct
  attribute :earthquakes_type, Types::String
  attribute :metadata,         Metadata
  attribute :features,         Types.Array(Feature)
  attribute :bbox,             Types.Array(Types::Double)

  def self.from_dynamic!(d)
    d = Types::Hash[d]
    new(
      earthquakes_type: d.fetch("type"),
      metadata:         Metadata.from_dynamic!(d.fetch("metadata")),
      features:         d.fetch("features").map { |x| Feature.from_dynamic!(x) },
      bbox:             d.fetch("bbox"),
    )
  end

  def self.from_json!(json)
    from_dynamic!(JSON.parse(json))
  end

  def to_dynamic
    {
      "type"     => @earthquakes_type,
      "metadata" => @metadata.to_dynamic,
      "features" => @features.map { |x| x.to_dynamic },
      "bbox"     => @bbox,
    }
  end

  def to_json(options = nil)
    JSON.generate(to_dynamic, options)
  end
end

File: dart/earthquakes.dart
// To parse this JSON data, do
//
//     final earthquakes = earthquakesFromJson(jsonString);

import 'dart:convert';

Earthquakes earthquakesFromJson(String str) {
    final jsonData = json.decode(str);
    return Earthquakes.fromJson(jsonData);
}

String earthquakesToJson(Earthquakes data) {
    final dyn = data.toJson();
    return json.encode(dyn);
}

class Earthquakes {
    String type;
    Metadata metadata;
    List<Feature> features;
    List<double> bbox;

    Earthquakes({
        this.type,
        this.metadata,
        this.features,
        this.bbox,
    });

    factory Earthquakes.fromJson(Map<String, dynamic> json) => new Earthquakes(
        type: json["type"],
        metadata: Metadata.fromJson(json["metadata"]),
        features: new List<Feature>.from(json["features"].map((x) => Feature.fromJson(x))),
        bbox: new List<double>.from(json["bbox"].map((x) => x.toDouble())),
    );

    Map<String, dynamic> toJson() => {
        "type": type,
        "metadata": metadata.toJson(),
        "features": new List<dynamic>.from(features.map((x) => x.toJson())),
        "bbox": new List<dynamic>.from(bbox.map((x) => x)),
    };
}

class Feature {
    String type;
    Properties properties;
    Geometry geometry;
    String id;

    Feature({
        this.type,
        this.properties,
        this.geometry,
        this.id,
    });

    factory Feature.fromJson(Map<String, dynamic> json) => new Feature(
        type: json["type"],
        properties: Properties.fromJson(json["properties"]),
        geometry: Geometry.fromJson(json["geometry"]),
        id: json["id"],
    );

    Map<String, dynamic> toJson() => {
        "type": type,
        "properties": properties.toJson(),
        "geometry": geometry.toJson(),
        "id": id,
    };
}

class Geometry {
    String type;
    List<double> coordinates;

    Geometry({
        this.type,
        this.coordinates,
    });

    factory Geometry.fromJson(Map<String, dynamic> json) => new Geometry(
        type: json["type"],
        coordinates: new List<double>.from(json["coordinates"].map((x) => x.toDouble())),
    );

    Map<String, dynamic> toJson() => {
        "type": type,
        "coordinates": new List<dynamic>.from(coordinates.map((x) => x)),
    };
}

class Properties {
    double mag;
    String place;
    int time;
    int updated;
    int tz;
    String url;
    String detail;
    int felt;
    int cdi;
    double mmi;
    dynamic alert;
    String status;
    int tsunami;
    int sig;
    String net;
    String code;
    String ids;
    String sources;
    String types;
    int nst;
    double dmin;
    double rms;
    int gap;
    String magType;
    String type;
    String title;

    Properties({
        this.mag,
        this.place,
        this.time,
        this.updated,
        this.tz,
        this.url,
        this.detail,
        this.felt,
        this.cdi,
        this.mmi,
        this.alert,
        this.status,
        this.tsunami,
        this.sig,
        this.net,
        this.code,
        this.ids,
        this.sources,
        this.types,
        this.nst,
        this.dmin,
        this.rms,
        this.gap,
        this.magType,
        this.type,
        this.title,
    });

    factory Properties.fromJson(Map<String, dynamic> json) => new Properties(
        mag: json["mag"].toDouble(),
        place: json["place"],
        time: json["time"],
        updated: json["updated"],
        tz: json["tz"],
        url: json["url"],
        detail: json["detail"],
        felt: json["felt"] == null ? null : json["felt"],
        cdi: json["cdi"] == null ? null : json["cdi"],
        mmi: json["mmi"] == null ? null : json["mmi"].toDouble(),
        alert: json["alert"],
        status: json["status"],
        tsunami: json["tsunami"],
        sig: json["sig"],
        net: json["net"],
        code: json["code"],
        ids: json["ids"],
        sources: json["sources"],
        types: json["types"],
        nst: json["nst"] == null ? null : json["nst"],
        dmin: json["dmin"] == null ? null : json["dmin"].toDouble(),
        rms: json["rms"].toDouble(),
        gap: json["gap"] == null ? null : json["gap"],
        magType: json["magType"],
        type: json["type"],
        title: json["title"],
    );

    Map<String, dynamic> toJson() => {
        "mag": mag,
        "place": place,
        "time": time,
        "updated": updated,
        "tz": tz,
        "url": url,
        "detail": detail,
        "felt": felt == null ? null : felt,
        "cdi": cdi == null ? null : cdi,
        "mmi": mmi == null ? null : mmi,
        "alert": alert,
        "status": status,
        "tsunami": tsunami,
        "sig": sig,
        "net": net,
        "code": code,
        "ids": ids,
        "sources": sources,
        "types": types,
        "nst": nst == null ? null : nst,
        "dmin": dmin == null ? null : dmin,
        "rms": rms,
        "gap": gap == null ? null : gap,
        "magType": magType,
        "type": type,
        "title": title,
    };
}

class Metadata {
    int generated;
    String url;
    String title;
    int status;
    String api;
    int count;

    Metadata({
        this.generated,
        this.url,
        this.title,
        this.status,
        this.api,
        this.count,
    });

    factory Metadata.fromJson(Map<String, dynamic> json) => new Metadata(
        generated: json["generated"],
        url: json["url"],
        title: json["title"],
        status: json["status"],
        api: json["api"],
        count: json["count"],
    );

    Map<String, dynamic> toJson() => {
        "generated": generated,
        "url": url,
        "title": title,
        "status": status,
        "api": api,
        "count": count,
    };
}
